File: Algorithm/Java/047. Permutations II.java
/**
* Author: WuLC
* Date:   2016-12-13 21:06:33
* Last modified by:   WuLC
* Last Modified time: 2016-12-13 21:07:42
* Email: liangchaowu5@gmail.com
*/


// recursive, same idea as 46. Permutations, but use a set to avoid duplicate number to be moved to the first place
public class Solution {
    
    private List<List<Integer>> result;
    public List<List<Integer>> permuteUnique(int[] nums) 
    {
        result = new ArrayList<List<Integer>>();
        dfs(nums, 0);
        return result;
    }
    
    public void dfs(int[] nums, int idx)
    {
        if (idx == nums.length)
        {
            List<Integer> tmp = new ArrayList<Integer>();
            for (int num:nums) tmp.add(num);
            result.add(tmp);
            return;
        }
        Set<Integer> appear = new HashSet<Integer>();
        for(int i=idx; i < nums.length; i++)
        {
            if (!appear.contains(nums[i]))
            {
                appear.add(nums[i]);
                swap(nums, idx, i);
                dfs(nums, idx+1);
                swap(nums, idx, i);
            }
        }
    }
    
    public void swap(int[] nums, int i, int j)
    {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}

File: Algorithm/Python/047. Permutations II.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-05-06 16:13:16
# @Last modified by:   WuLC
# @Last Modified time: 2016-05-06 16:15:29
# @Email: liangchaowu5@gmail.com

# 直接进行排列，注意从后往前插入数字，从前往后会出错
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = [[]]
        for num in nums:
            new_result = []
            for seq in result:
                n = len(seq)
                for i in xrange(n,-1,-1):
                    if i < n and seq[i] == num:
                        break
                    new_result.append(seq[:i]+[num]+seq[i:])
            result = new_result
        return result
                        
# divide and conquer
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        return self.helper(nums)
    
    def helper(self, nums):
        result = []
        if len(nums) == 1:
            result.append(nums)
        else:
            for i in xrange(len(nums)):
                if i > 0 and nums[i] == nums[i-1]:
                    continue
                else:
                    for left in self.helper(nums[:i]+nums[i+1:]):
                        result.append([nums[i]]+left)
        return result